grafana/grafana
# Loki: Add feature flag to enable dataplane-compliant metric frames

this PR adds a feature-flag that makes Loki's responses to metric queries compliant with the dataplane-spec's [time series multi]( https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/contract_docs/timeseries.md#time-series-multi-format-timeseriesmulti ) and [numeric multi]( https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/contract_docs/numeric.md#numeric-multi-format-numericmulti ) formats.

(similar PR happened to Prometheus in https://github.com/grafana/grafana/pull/62694)

this means the following changes:
- the `Frame.Name` field is not used anymore
- for "vector" data (numbers for a single timestamp), the `Frame.Meta.Type` field changes from `timeseries-multi` to `numeric-multi`.

we do not expect many backward-compatibility issues with this change, but for now it is opt-in thing controlled by the feature-flag.

we are running the snapshot-tests twice, both for the flag being enabled and also for the flag being disabled

when reading the diff, the real changes are in the `tsdb/loki` part, the rest is adding a feature-flag, and adding more snapshot-tests (because now we need to cover both flag-enabled and flag-disabled cases).

if anyone has any idea how to do this in an easier way, i'm listening 👍 .


how to test:
1. enable the feature-flag `lokiMetricDataplane`
2. go to explore, run a metric LogQL query, make sure the graph you see looks ok
3. repeat the same in a dashboard
4. now disable the feature-flag `lokiMetricDataplane`
5. repeat the tests from [2] and [3]
